### DISCUSSION TITLE
Listing directories

### DISCUSSION THREAD

**Discussion by Radio-biscuit** [savonet/liquidsoap#4038] [Q&A] [answered]
Hello,

I've searched for some documentation and discussions on this, but I can't find anything that answers my question:

I want to get a list of directories from file.ls. It should be fairly easy by using list.iter and file.is_directory. I can get the directory entry, but I want to add this to a list and I keep getting type errors when I use list.add(directory, list). How can I get around this? I was making the target list a string, but maybe that's where the trouble is? If I use an empty list to start, the same problem occurs.

Any help would be great,

Thanks in advance! A.

**Comment by toots** [comment]
Without looking at your code it's hard to give more help!

**Reply by Radio-biscuit** [reply]
Sorry, yes I realised after.

The code is:
```
o=[]
f=file.ls("/directory')
list.iter(fun(x)->if file.is_directory(x) then o=list.add(x,o) end, f)
```

I'm now getting unused variable "o" instead. Before I was getting type errors "[] is not ()".

Thanks again. A.

**Reply by toots** [reply, marked as answer]
So, `list.iter` does not produce any result, it justs applies the function given as argument and discards its result.

You can do one of two things:
1. Direct filter

This is the most straight forward:
```liquidsoap
directories = list.filter(
  file.is_directory,
  file.ls(absolute=true, "/directory")
)
```

2. Use a reducer/fold

This is the most advanced/functional way. Hard to understand at first but very powerful. You can use `list.fold` for that, it accumulates the result of applying the reduce function over and over:
```liquidsoap
basedir = "/directory"

directories = list.fold(
  fun (directories, f) -> begin
    f = path.concat(basedir, f)
    if file.is_directory(f) then
      [f, ...directories]
    else
      directories
    end
  end,
  [],
  file.ls(basedir),
)
```

**Reply by Radio-biscuit** [reply]
That explains a lot! I'm more conversant with C++, so I look at it like "it goes out of scope" so there's no result. The fold technique was my next attempt, and I didn't realise list.filter could work that way too. I have options now!

Thank you for the help (from a noobie coder). A.

**Reply by toots** [reply]
You're welcome!
